### ISSUE TITLE
`nullified` fields in consecutive `/export` responses are "incoherent"

### ISSUE THREAD

**Issue by DavidBuchanan314** [did-method-plc/did-method-plc#56]
IIUC, the `nullified` field indicates an operation that has been subsequently rolled back due to being "trumped" by a higher-ranking rotation key, within the 72h window.

It is therefore a field whose value can change over time. If you're looking at a set of multiple operations for a given DID, their respective `nullified` field values only have coherent values/meanings if all operations were sampled at the same instant in time.

For example, if you're looking at a response from `https://plc.directory/<did>/log/audit`, then all `nullified` fields that you see were sampled at the same time instant (or equivalent, using db transactions/locking/whatever (I hope!)) and therefore you can safely make sense of their values. Good!

However, if you've been enumerating operations via the `/export` endpoint, you've inevitably been sampling the data from multiple points in time, and by the time you've got to the end, there's a chance some of your `nullified` fields are stale.

I don't think there's any practical way to make use of the `nullified` fields here - you have to look at the operations you have, in order[^1], and their `createdAt` timestamps, and run the logic to decide for yourself whether they've been nullified. Or am I overthinking things? I think these nullifications will be relatively rare, but when they do happen it's perhaps a contentious situation, and therefore all the more important that observers can correctly come to a consensus about what happened.

[^1]: Presumably, two operations could have identical `createdAt` timestamps. What happens then?

As such, I think it would be safer to just not return the `nullified` field in `/export` responses, to save clients from making bad assumptions. As a bonus, your responses become more cacheable!

**Comment by bnewbold** [comment]
Yeah, that makes sense. We probably should not populate `nullified` with either value in the `/export` stream (but of course should continue to include `createdAt`).

**Comment by str4d** [comment]
I'm just at the point of dealing with this issue in my PLC mirror (until now I've just imported the `nullified` field from the source as-is).

- On the one hand, I agree that it changing over time means that in the steady state, exporting it to other mirrors is not good as it cannot be relied upon.
- On the other hand, any exported entries with a `createdAt` that is more than 72 hours in the past _should_ have stable `nullified` fields, and in fact it would be a major problem if those `nullified` field values changed post-export (indicating a consensus divergence). So having the values present in exports would enable the mirrors to check the source's work during import.
- On the first hand again, having it present anywhere in the export means the importers are less likely to realise they need to do these checks, and having it only present when stable makes the export API more complex (as `nullified` would itself then be a nullable field).

The overhead of checking every new record in the steady state should be fine (IIRC a few days ago I was observing around 1 operation per second when running my mirror). However, doing this for all ~14 million entries (at time of writing) individually as the mirror is built from scratch is likely to introduce significant overhead (due to the random-access lookup of the existing audit log for whichever DID each operation happens to be for), and slow down an already slow import process (which necessarily has to be single-threaded and sequential in order to perform these checks).

I'd be more inclined to perform a direct import of all entries first, and then run an audit over the entire mirror's contents post-import (at which point we have all of the records for each DID, and can check their audit logs independently in parallel).

- If the source continues to provide `nullified` fields, then I'd trust them during the direct import, and then check them during the audit.
  - We could optimistically derive `nullified` fields near the chain tip if the export stops including them; if we detect a failure during the audit phase then we'd a) correct everything after that, and b) alert that the source has a consensus divergence.
- If the `nullified` fields stop being populated, then I'd stick nulls into the database during the direct import, and then fill them in with the correct values during the audit.
  - We cannot derive `nullified` fields near the chain tip before completing the audit in this case, as we cannot know what past operations are active.

Thinking aloud, maybe what mirrors actually need here is two modes, that the operator can toggle between depending on performance requirements:
- A "catch-up" mode that imports without per-operation nullified checks, and then audits/prepares in a single pass.
- A "run" mode that requires all `nullified` fields to be set for all entries (implying that the "catch-up" ran to completion), and does the `nullified` checks for every imported entry.

I'll experiment with my PLC mirror and see what the performance costs are like.

**Comment by DavidBuchanan314** [comment]
Checking for conflicting operations (i.e. a second op with the same `prev`) should be an O(logn) with typical db indexing, or even O(1) if you get fancy (hashtable, bloom filter, etc) - I don't think auditing all 14m entries should take long at all. I believe signature verification time would be the overall bottleneck by far.

A while back I wrote some code to just look for nullified ops and I don't remember it taking long to run (probably on the order of minutes with python+sqlite)

**Comment by DavidBuchanan314** [comment]
You can parallelize things as long as you don't have two ops for the same DID being processed in parallel
